Create confirmedplace.py

sebastianbarry opened the pull request:
To create confirmedplace.py, I also created expectedplace.py and inferredplace.py

@shankari I have not tested this by clearing and running the pipeline yet. I believe I still have other places to add calls to confirmedplace (like maybe `emission/core/wrapper/entry.py`??)

sebastianbarry committed:
Create confirmedplace.py

To create confirmedplace.py, I also created expectedplace.py and inferredplace.py

shankari commented:
@sebastianbarry 

> I believe I still have other places to add calls to confirmedplace (like maybe emission/core/wrapper/entry.py??)

did you look at the details related to wrapper classes in the e-mission documentation? And chapter 5 of my thesis?

> I also created expectedplace.py and inferredplace.py

I am not sure we need `inferredplaced` and `expectedplace` since we don't currently use label assist algorithms for places, but we can discuss that at our next meeting.

sebastianbarry commented:
Here is the guide for adding a new data type to the server: https://github.com/e-mission/e-mission-docs/blob/4fad0c4dd82893529bdc808fc5a8329f98ea7367/docs/dev/back/adding_a_new_data_type.md

It says:

> * include a reference to it in entry.py
> * in builtin_timeseries.py
> * add formatters that can expand fields if necessary (e.g. https://github.com/e-mission/e-mission-server/blob/master/emission/net/usercache/formatters/android/location.py)

shankari commented:
> I am not sure we need inferredplaced and expectedplace since we don't currently use label assist algorithms for places, but we can discuss that at our next meeting.

So we will need these if we ever start inferring labels for places. We don't even *have* labels for places at this point, and it is not clear what those labels might be and how well we will be able to infer them in the future, so I think we should not overengineer this and we should not include inferred place and expected place for now. We can include them in the future once we have a concrete use case.

sebastianbarry committed:
Added confirmedplace in entry.py and builtin_timeseries.py

sebastianbarry committed:
Removed expectedplace.py and inferredplace.py

Because these files are not needed as places do not need inferred and expected, i am removing them

shankari commented:
Prior intern PR to add a stage to the pipeline https://github.com/e-mission/e-mission-server/pull/825

You can also look up other PRs by the same intern in case you need to understand the label inference mechanism in greater detail

Note that Gabriel also added a UI proposal that gives you some [additional context around inferred labels](https://objects.githubusercontent.com/github-production-repository-file-5c1aeb/134500339/6743130?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230213%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230213T165234Z&X-Amz-Expires=300&X-Amz-Signature=74c62e0fff122f6594e2282ca9cc93514108feb716d1b267a6b586b39297206d&X-Amz-SignedHeaders=host&actor_id=2423263&key_id=0&repo_id=134500339&response-content-disposition=attachment%3Bfilename%3DUI.Draft.2.pdf&response-content-type=application%2Fpdf)

sebastianbarry committed:
Added confirmedplace into the pipeline

pipelinestate.py
- Added a 17th ENUM to the PipelineStages named "CREATE_PLACE_OBJECTS"

reset.py
- Added a Pipeline stage to the stages_with_fuzz array for CREATE_PLACE_OBJECTS

common.py
- Added a "create_place_objects" function after all the other pipeline stage functions

intake_stage.py
- Added a stage for the intake to store the pipeline time for CREATE_PLACE_OBJECTS, and call the create_place_object function

confirmedplace.py
- Removed expected_place and primary_section because @shankari  said in our last meeting that the confirmedplace object does not need these

matcher.py
- Created the create_place_objects function which follows the same format as create_confirmed_objects, but only for trips. It uses the same existing mark_confirmed_object_creation_done/failed functions, and the same get_time_range_for_confirmed_object_creation
- Added a create_confirmed_places function which follows the same format as create_confirmed_trips; I removed any data which does not exist in the confirmed_places emission/core/wrapper/confirmedplace.py data model

shankari commented on emission/core/wrapper/confirmedplace.py:
```diff
@@ -15,10 +15,8 @@ class Confirmedplace(ecwp.Place):
                   "inferred_labels": ecwb.WrapperBase.Access.WORM,
                   "expectation": ecwb.WrapperBase.Access.WORM,
```
note that we are not supporting place-level inferences right now, so no need to include these either.

shankari commented on emission/core/wrapper/pipelinestate.py:
```diff
@@ -27,6 +27,7 @@ class PipelineStages(enum.Enum):
     USER_MODEL = 7
     RECOMMENDATION = 8
     OUTPUT_GEN = 9
+    CREATE_PLACE_OBJECTS = 17
```
there is already a `CREATE_CONFIRMED_OBJECTS`. This is a confirmed object. So why do we need to create a new stage for it?

shankari commented on emission/pipeline/intake_stage.py:
```diff
@@ -184,6 +184,14 @@ def run_intake_pipeline_for_user(uuid):
         esds.store_pipeline_time(uuid, ecwp.PipelineStages.CREATE_CONFIRMED_OBJECTS.name,
                                  time.time(), crt.elapsed)
 
+        with ect.Timer() as crt:
+            logging.info("*" * 10 + "UUID %s: creating place objects " % uuid + "*" * 10)
+            print(str(arrow.now()) + "*" * 10 + "UUID %s: creating place objects " % uuid + "*" * 10)
+            eaum.create_place_objects(uuid)
+
+        esds.store_pipeline_time(uuid, ecwp.PipelineStages.CREATE_PLACE_OBJECTS.name,
+                                 time.time(), crt.elapsed)
```
Ditto.

shankari commented on emission/pipeline/reset.py:
```diff
@@ -187,7 +187,8 @@ def reset_pipeline_state(user_id, reset_ts, is_dry_run):
                         ecwp.PipelineStages.EXPECTATION_POPULATION.value,
-                        ecwp.PipelineStages.CREATE_CONFIRMED_OBJECTS.value]
+                        ecwp.PipelineStages.CREATE_CONFIRMED_OBJECTS.value,
+                        ecwp.PipelineStages.CREATE_PLACE_OBJECTS.value]
```
Ditto.

shankari commented on emission/tests/common.py:
```diff
@@ -195,6 +195,7 @@ def runIntakePipeline(uuid):
     eacilp.infer_labels(uuid)
     eaue.populate_expectations(uuid)
     eaum.create_confirmed_objects(uuid)
+    eaum.create_place_objects(uuid)
```
Ditto.

shankari commented on emission/analysis/userinput/matcher.py:
```diff
@@ -87,7 +87,48 @@ def create_confirmed_objects(user_id):
+        logging.exception("Error while creating place objects, timestamp is unchanged")
+        epq.mark_confirmed_object_creation_failed(user_id)
+
+def create_confirmed_places(user_id, timerange):
```
This can be called directly from the existing pipeline step.

shankari reviewed (changes requested):
This is a not-too-bad start but it is not clear to me that you need to add a new pipeline stage here instead of using an existing one.
What do we do for existing trip/place creation, for example? Do we have separate steps?
Also, I anticipate that this will not work; have you tried to run the pipeline with it enabled?

JGreenlee commented:
For visibility, I'm posting a recap of recent progress here since I won't be online tomorrow

On wednesday, @sebastianbarry and I started testing things out on the [add-composite-trip](https://github.com/sebastianbarry/e-mission-server/tree/add-composite-trip) branch
we had already decided to incorporate confirmed place creation into the existing pipleline stage instead of it being its own stage

wednesday night I _did_ add a new stage, which creates composite trips. It works, but needs review because I don't believe I did it in the cleanest or most desirable way (I just wanted to get it working for a first draft) The structure of composite trip may also be reconsidered.
(This composite trip creation _could_ just happen at the end of the confirmed objects creation stage, but I thought since "composite" is representing a new class of object for the presentation layer, it warranted a new pipeline stage. is that right?)

Maybe we will want to make a new PR from the [add-composite-trip](https://github.com/sebastianbarry/e-mission-server/tree/add-composite-trip) branch instead, or maybe we just want to merge those changes into this branch
I will let @sebastianbarry organize the PR(s) however you see fit

sebastianbarry commented:
Added "add composite trip", closing this Pull Request in favor of that one

shankari commented:
>  we had already decided to incorporate confirmed place creation into the existing pipleline stage instead of it being its own stage

Great! So I will check the current version for that

> wednesday night I did add a new stage, which creates composite trips. (This composite trip creation could just happen at the end of the confirmed objects creation stage, but I thought since "composite" is representing a new class of object for the presentation layer, it warranted a new pipeline stage. is that right?)

Yes, your intuition is correct at the high level. We do want to have composite trip created in a separate pipeline step. Note that if we change the retrieval method back to retrieving the components of composite trip separately, then we can just remove that pipeline step. Keeping it as a separate step reinforces that separation between the data model and the data presentation.
